### ISSUE TITLE
mbedTLS  3.6 handshake failure (40) during TLS negotiation

### ISSUE THREAD

**Issue by roehling** [libgit2/libgit2#6950]
### Reproduction steps
Build with mbedTLS and clone an arbitrary repository from GitHub

### Expected behavior
The repository is cloned

### Actual behavior
`SSL error: 0xffff8880 - SSL - A fatal alert message was received from our peer`

### Version of libgit2 (release number or SHA1)
1.8.4

### Operating system(s) tested
Debian

**Comment by roehling** [comment]
I have traced the issue to the cipher list provided in `src/libgit2/streams/mbedtls.c`. The TLS connection works as expected if the list of allowed ciphers is not overridden:

```diff
--- a/src/libgit2/streams/mbedtls.c
+++ b/src/libgit2/streams/mbedtls.c
@@ -104,6 +104,7 @@ int git_mbedtls_stream_global_init(void)
         * is made, so we can never see the certificate and override it. */
        mbedtls_ssl_conf_authmode(&mbedtls_config, MBEDTLS_SSL_VERIFY_OPTIONAL);

+#if 0
        /* set the list of allowed ciphersuites */
        ciphers_known = 0;
        cipher_string = cipher_string_tmp = git__strdup(GIT_SSL_DEFAULT_CIPHERS);
@@ -127,6 +128,7 @@ int git_mbedtls_stream_global_init(void)
                goto cleanup;
        }
        mbedtls_ssl_conf_ciphersuites(&mbedtls_config, ciphers_list);
+#endif

        /* Seeding the random number generator */
```

Is there a reason why the cipher list is explicitly set? AFAICT the forced ciphers are not even the most modern ones, and the mbedTLS library seems perfectly capable of selecting properly secure ciphers on its own.

**Comment by ethomson** [comment]
I don't know what the rationale is here, to be honest. The commit messages may be elucidating but this is probably something that we've been doing for a long time.

I agree that it seems like something we shouldn't be doing unless we actually know that the underlying TLS engine is doing a bad job, _and_ I don't think that's something that we (meaning libgit2 contributors) have the expertise in to be able to say ourselves.

I see a few options:

1. Update this to a set of good default recommendations that some other organization with expertise has produced. (eg NIST) And, crucially, build a mechanism to remind us to update this every n months. This reminder is crucial because there's a very real possibility that we are having this conversation because we did the first part of this option once and then forgot about it.

2. Drop this but make sure that it remains user configurable. That way if a user knows that their version of some TLS library has bad defaults but can't update it, they can set the ciphers. Is this actually a thing though?

**Comment by roehling** [comment]
> That way if a user knows that their version of some TLS library has bad defaults but can't update it, they can set the ciphers. Is this actually a thing though?

I've seen *server* admins tweak the cipher suite; back in the TLS 1.1 days to enforce forward secrecy when some clients would prefer cipher suites with non-ephemeral key exchanges, and again when RC4 was broken. *Clients* usually do things like certificate pinning if they only ever talk to one particular service (e.g. the GMail app talking to the GMail servers), but I don't think this applies to a library like libgit2 which is supposed to talk to arbitrary Git servers.

Personally, I would say it should be the responsibility of the server side to reject unsafe cipher suites, as that will benefit all clients at once. Updating thousands or millions of clients individually takes ages, and for some clients such as IoT devices, it may never happen at all.

Also, I would trust the TLS library authors to provide sane defaults and not make this user-configurable, because for every expert user who can actually make good use of this option, there will be a thousand cargo cultists who just blindly copy a random configuration from Stack Overflow.

I'm not a cryptography expert though, so take my opinion with the appropriate grain of salt.

**Comment by ethomson** [comment]
Yes, I think that your analysis is correct. It would be worth digging through history to see if we can understand the rationale a bit. There may be some gotcha here (OpenSSL, at least, has been a bit shit at times in its history).

**Comment by ethomson** [comment]
Looking back in history, it seems like this was added because OpenSSL had bad defaults... https://github.com/libgit2/libgit2/pull/3683

So I think that we should understand: did they _really_ have bad defaults? Do they _still_ have bad defaults that we should have an opinion about? If not, do we still support versions with bad defaults?

If we think that OpenSSL is "wrong" here, then perhaps the wiser thing to do here is to have a deny list for the things that we want to _exclude_.

**Comment by ethomson** [comment]
## OpenSSL

Right, so. I just compiled openssl 3.4.0, and here's the list of ciphers that are supported out-of-the-box:

```
morbo:~% openssl ciphers
TLS_AES_256_GCM_SHA384:TLS_CHACHA20_POLY1305_SHA256:TLS_AES_128_GCM_SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:DHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:DHE-RSA-CHACHA20-POLY1305:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES256-SHA384:DHE-RSA-AES256-SHA256:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA256:DHE-RSA-AES128-SHA256:ECDHE-ECDSA-AES256-SHA:ECDHE-RSA-AES256-SHA:DHE-RSA-AES256-SHA:ECDHE-ECDSA-AES128-SHA:ECDHE-RSA-AES128-SHA:DHE-RSA-AES128-SHA:RSA-PSK-AES256-GCM-SHA384:DHE-PSK-AES256-GCM-SHA384:RSA-PSK-CHACHA20-POLY1305:DHE-PSK-CHACHA20-POLY1305:ECDHE-PSK-CHACHA20-POLY1305:AES256-GCM-SHA384:PSK-AES256-GCM-SHA384:PSK-CHACHA20-POLY1305:RSA-PSK-AES128-GCM-SHA256:DHE-PSK-AES128-GCM-SHA256:AES128-GCM-SHA256:PSK-AES128-GCM-SHA256:AES256-SHA256:AES128-SHA256:ECDHE-PSK-AES256-CBC-SHA384:ECDHE-PSK-AES256-CBC-SHA:SRP-RSA-AES-256-CBC-SHA:SRP-AES-256-CBC-SHA:RSA-PSK-AES256-CBC-SHA384:DHE-PSK-AES256-CBC-SHA384:RSA-PSK-AES256-CBC-SHA:DHE-PSK-AES256-CBC-SHA:AES256-SHA:PSK-AES256-CBC-SHA384:PSK-AES256-CBC-SHA:ECDHE-PSK-AES128-CBC-SHA256:ECDHE-PSK-AES128-CBC-SHA:SRP-RSA-AES-128-CBC-SHA:SRP-AES-128-CBC-SHA:RSA-PSK-AES128-CBC-SHA256:DHE-PSK-AES128-CBC-SHA256:RSA-PSK-AES128-CBC-SHA:DHE-PSK-AES128-CBC-SHA:AES128-SHA:PSK-AES128-CBC-SHA256:PSK-AES128-CBC-SHA
```

Holy crow, that's a lot. The last on the list is `PSK-AES128-CBC-SHA` which is [weak](https://ciphersuite.info/cs/TLS_PSK_WITH_AES_128_CBC_SHA/).

And, GitHub blocks it themselves:

```
morbo:~% openssl s_client -cipher TLS-RSA-WITH-AES-256-CBC-SHA -connect www.github.com:443
Call to SSL_CONF_cmd(-cipher, TLS-RSA-WITH-AES-256-CBC-SHA) failed
400FCD0302000000:error:0A0000B9:SSL routines:SSL_CTX_set_cipher_list:no cipher match:ssl/ssl_lib.c:3346:
```

So, that's nice _if you're connecting to GitHub_ but I think that we should continue to be proactive. Here's the [Mozilla _server side_ TLS list](https://wiki.mozilla.org/Security/Server_Side_TLS) for intermediate security:

```
ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY13    05:ECDHE-RSA-CHACHA20-POLY1305:DHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384:DHE-RSA-CHACHA20-POLY1305
```

Which seems like a reasonable improvement over what we have today. (We include things like `TLS-RSA-WITH-AES-256-CBC-SHA` which - hey - also [weak](https://ciphersuite.info/cs/TLS_RSA_WITH_AES_256_CBC_SHA256/).)

So it seems useful to continue to limit the ciphers on OpenSSL.

## mbedTLS

Similarly, looking at mbedTLS, the supported cipher list out of the box is _huge_:

```
hypnotoad:mbedtls-3.5.2/programs/ssl% ./ssl_client2

 usage: ssl_client2 param=<>...

 acceptable parameters:
    server_name=%s      default: localhost
    server_addr=%s      default: given by name
    server_port=%d      default: 4433
    request_page=%s     default: "."
    request_size=%d     default: about 34 (basic request)
                        (minimum: 0, max: 20000)
                        If 0, in the first exchange only an empty
                        application data message is sent followed by
                        a second non-empty message before attempting
                        to read a response from the server
    debug_level=%d      default: 0 (disabled)
    build_version=%d    default: none (disabled)
                        option: 1 (print build version only and stop)
    nbio=%d             default: 0 (blocking I/O)
                        options: 1 (non-blocking), 2 (added delays)
    event=%d            default: 0 (loop)
                        options: 1 (level-triggered, implies nbio=1),
    read_timeout=%d     default: 0 ms (no timeout)
    max_resend=%d       default: 0 (no resend on timeout)
    skip_close_notify=%d default: 0 (send close_notify)

    dtls=%d             default: 0 (TLS)
    hs_timeout=%d-%d    default: (library default: 1000-60000)
                        range of DTLS handshake timeouts in millisecs
    mtu=%d              default: (library default: unlimited)
    dgram_packing=%d    default: 1 (allowed)
                        allow or forbid packing of multiple
                        records within a single datgram.
    cid=%d             Disable (0) or enable (1) the use of the DTLS Connection ID extension.
                       default: 0 (disabled)
    cid_renego=%d      Disable (0) or enable (1) the use of the DTLS Connection ID extension during renegotiation.
                       default: same as 'cid' parameter
    cid_val=%s          The CID to use for incoming messages (in hex, without 0x).
                        default: ""
    cid_val_renego=%s   The CID to use for incoming messages (in hex, without 0x) after renegotiation.
                        default: same as 'cid_val' parameter

    auth_mode=%s        default: (library default: none)
                        options: none, optional, required
    ca_file=%s          The single file containing the top-level CA(s) you fully trust
                        default: "" (pre-loaded)
                        use "none" to skip loading any top-level CAs.
    ca_path=%s          The path containing the top-level CA(s) you fully trust
                        default: "" (pre-loaded) (overrides ca_file)
                        use "none" to skip loading any top-level CAs.
    crt_file=%s         Your own cert and chain (in bottom to top order, top may be omitted)
                        default: "" (pre-loaded)
    key_file=%s         default: "" (pre-loaded)
    key_pwd=%s          Password for key specified by key_file argument
                        default: none

    psk=%s              default: "" (disabled)
                          The PSK values are in hex, without 0x.
    psk_identity=%s     default: "Client_identity"

    allow_legacy=%d     default: (library default: no)
    renegotiation=%d    default: 0 (disabled)
    renegotiate=%d      default: 0 (disabled)
    exchanges=%d        default: 1
    reconnect=%d        number of reconnections using session resumption
                        default: 0 (disabled)
    reco_server_name=%s  default: NULL
    reco_delay=%d       default: 0 milliseconds
    reco_mode=%d        0: copy session, 1: serialize session
                        default: 1
    reconnect_hard=%d   default: 0 (disabled)
    tickets=%d          default: 1 (enabled)
    eap_tls=%d          default: 0 (disabled)
    max_frag_len=%d     default: 16384 (tls default)
                        options: 512, 1024, 2048, 4096
    context_crt_cb=%d   This determines whether the CRT verification callback is bound
                        to the SSL configuration of the SSL context.
                        Possible values:
                        - 0 (default): Use CRT callback bound to configuration
                        - 1: Use CRT callback bound to SSL context
    alpn=%s             default: "" (disabled)
                        example: spdy/1,http/1.1
    extended_ms=0/1     default: (library default: on)
    etm=0/1             default: (library default: on)
    reproducible=0/1     default: 0 (disabled)
    groups=a,b,c,d      default: "default" (library default)
                        example: "secp521r1,brainpoolP512r1"
                        - use "none" for empty list
                        - see mbedtls_ecp_curve_list()
                          for acceptable EC group names
                        - the following ffdh groups are supported:
                          ffdhe2048, ffdhe3072, ffdhe4096, ffdhe6144,
                          ffdhe8192
    sig_algs=a,b,c,d      default: "default" (library default)
                          example: "ecdsa_secp256r1_sha256,ecdsa_secp384r1_sha384"
    dhmlen=%d           default: (library default: 1024 bits)
    key_opaque_algs=%s  Allowed opaque key algorithms.
                        comma-separated pair of values among the following:
                        rsa-sign-pkcs1, rsa-sign-pss, rsa-sign-pss-sha256,
                        rsa-sign-pss-sha384, rsa-sign-pss-sha512, rsa-decrypt,
                        ecdsa-sign, ecdh, none (only acceptable for
                        the second value).

    allow_sha1=%d       default: 0
    min_version=%s      default: (library default: tls12)
    max_version=%s      default: (library default: tls12)
    force_version=%s    default: "" (none)
                         options: tls12, dtls12

    force_ciphersuite=<name>    default: all enabled
    query_config=<name>         return 0 if the specified
                                configuration macro is defined and 1
                                otherwise. The expansion of the macro
                                is printed if it is defined
    serialize=%d        default: 0 (do not serialize/deserialize)
                        options: 1 (serialize)
                                 2 (serialize with re-initialization)
    context_file=%s     The file path to write a serialized connection
                        in the form of base64 code (serialize option
                        must be set)
                         default: "" (do nothing)
                         option: a file path
 acceptable ciphersuite names:
 TLS-ECDHE-RSA-WITH-CHACHA20-POLY1305-SHA256 TLS-ECDHE-ECDSA-WITH-CHACHA20-POLY1305-SHA256
 TLS-DHE-RSA-WITH-CHACHA20-POLY1305-SHA256  TLS-ECDHE-ECDSA-WITH-AES-256-GCM-SHA384
 TLS-ECDHE-RSA-WITH-AES-256-GCM-SHA384      TLS-DHE-RSA-WITH-AES-256-GCM-SHA384
 TLS-ECDHE-ECDSA-WITH-AES-256-CCM           TLS-DHE-RSA-WITH-AES-256-CCM
 TLS-ECDHE-ECDSA-WITH-AES-256-CBC-SHA384    TLS-ECDHE-RSA-WITH-AES-256-CBC-SHA384
 TLS-DHE-RSA-WITH-AES-256-CBC-SHA256        TLS-ECDHE-ECDSA-WITH-AES-256-CBC-SHA
 TLS-ECDHE-RSA-WITH-AES-256-CBC-SHA         TLS-DHE-RSA-WITH-AES-256-CBC-SHA
 TLS-ECDHE-ECDSA-WITH-AES-256-CCM-8         TLS-DHE-RSA-WITH-AES-256-CCM-8
 TLS-ECDHE-ECDSA-WITH-CAMELLIA-256-GCM-SHA384 TLS-ECDHE-RSA-WITH-CAMELLIA-256-GCM-SHA384
 TLS-DHE-RSA-WITH-CAMELLIA-256-GCM-SHA384   TLS-ECDHE-ECDSA-WITH-CAMELLIA-256-CBC-SHA384
 TLS-ECDHE-RSA-WITH-CAMELLIA-256-CBC-SHA384 TLS-DHE-RSA-WITH-CAMELLIA-256-CBC-SHA256
 TLS-DHE-RSA-WITH-CAMELLIA-256-CBC-SHA      TLS-ECDHE-ECDSA-WITH-ARIA-256-GCM-SHA384
 TLS-ECDHE-RSA-WITH-ARIA-256-GCM-SHA384     TLS-DHE-RSA-WITH-ARIA-256-GCM-SHA384
 TLS-ECDHE-ECDSA-WITH-ARIA-256-CBC-SHA384   TLS-ECDHE-RSA-WITH-ARIA-256-CBC-SHA384
 TLS-DHE-RSA-WITH-ARIA-256-CBC-SHA384       TLS-ECDHE-ECDSA-WITH-AES-128-GCM-SHA256
 TLS-ECDHE-RSA-WITH-AES-128-GCM-SHA256      TLS-DHE-RSA-WITH-AES-128-GCM-SHA256
 TLS-ECDHE-ECDSA-WITH-AES-128-CCM           TLS-DHE-RSA-WITH-AES-128-CCM
 TLS-ECDHE-ECDSA-WITH-AES-128-CBC-SHA256    TLS-ECDHE-RSA-WITH-AES-128-CBC-SHA256
 TLS-DHE-RSA-WITH-AES-128-CBC-SHA256        TLS-ECDHE-ECDSA-WITH-AES-128-CBC-SHA
 TLS-ECDHE-RSA-WITH-AES-128-CBC-SHA         TLS-DHE-RSA-WITH-AES-128-CBC-SHA
 TLS-ECDHE-ECDSA-WITH-AES-128-CCM-8         TLS-DHE-RSA-WITH-AES-128-CCM-8
 TLS-ECDHE-ECDSA-WITH-CAMELLIA-128-GCM-SHA256 TLS-ECDHE-RSA-WITH-CAMELLIA-128-GCM-SHA256
 TLS-DHE-RSA-WITH-CAMELLIA-128-GCM-SHA256   TLS-ECDHE-ECDSA-WITH-CAMELLIA-128-CBC-SHA256
 TLS-ECDHE-RSA-WITH-CAMELLIA-128-CBC-SHA256 TLS-DHE-RSA-WITH-CAMELLIA-128-CBC-SHA256
 TLS-DHE-RSA-WITH-CAMELLIA-128-CBC-SHA      TLS-ECDHE-ECDSA-WITH-ARIA-128-GCM-SHA256
 TLS-ECDHE-RSA-WITH-ARIA-128-GCM-SHA256     TLS-DHE-RSA-WITH-ARIA-128-GCM-SHA256
 TLS-ECDHE-ECDSA-WITH-ARIA-128-CBC-SHA256   TLS-ECDHE-RSA-WITH-ARIA-128-CBC-SHA256
 TLS-DHE-RSA-WITH-ARIA-128-CBC-SHA256       TLS-ECDHE-PSK-WITH-CHACHA20-POLY1305-SHA256
 TLS-DHE-PSK-WITH-CHACHA20-POLY1305-SHA256  TLS-DHE-PSK-WITH-AES-256-GCM-SHA384
 TLS-DHE-PSK-WITH-AES-256-CCM               TLS-ECDHE-PSK-WITH-AES-256-CBC-SHA384
 TLS-DHE-PSK-WITH-AES-256-CBC-SHA384        TLS-ECDHE-PSK-WITH-AES-256-CBC-SHA
 TLS-DHE-PSK-WITH-AES-256-CBC-SHA           TLS-DHE-PSK-WITH-CAMELLIA-256-GCM-SHA384
 TLS-ECDHE-PSK-WITH-CAMELLIA-256-CBC-SHA384 TLS-DHE-PSK-WITH-CAMELLIA-256-CBC-SHA384
 TLS-DHE-PSK-WITH-AES-256-CCM-8             TLS-DHE-PSK-WITH-ARIA-256-GCM-SHA384
 TLS-ECDHE-PSK-WITH-ARIA-256-CBC-SHA384     TLS-DHE-PSK-WITH-ARIA-256-CBC-SHA384
 TLS-DHE-PSK-WITH-AES-128-GCM-SHA256        TLS-DHE-PSK-WITH-AES-128-CCM
 TLS-ECDHE-PSK-WITH-AES-128-CBC-SHA256      TLS-DHE-PSK-WITH-AES-128-CBC-SHA256
 TLS-ECDHE-PSK-WITH-AES-128-CBC-SHA         TLS-DHE-PSK-WITH-AES-128-CBC-SHA
 TLS-DHE-PSK-WITH-CAMELLIA-128-GCM-SHA256   TLS-DHE-PSK-WITH-CAMELLIA-128-CBC-SHA256
 TLS-ECDHE-PSK-WITH-CAMELLIA-128-CBC-SHA256 TLS-DHE-PSK-WITH-AES-128-CCM-8
 TLS-DHE-PSK-WITH-ARIA-128-GCM-SHA256       TLS-ECDHE-PSK-WITH-ARIA-128-CBC-SHA256
 TLS-DHE-PSK-WITH-ARIA-128-CBC-SHA256       TLS-RSA-WITH-AES-256-GCM-SHA384
 TLS-RSA-WITH-AES-256-CCM                   TLS-RSA-WITH-AES-256-CBC-SHA256
 TLS-RSA-WITH-AES-256-CBC-SHA               TLS-ECDH-RSA-WITH-AES-256-GCM-SHA384
 TLS-ECDH-RSA-WITH-AES-256-CBC-SHA384       TLS-ECDH-RSA-WITH-AES-256-CBC-SHA
 TLS-ECDH-ECDSA-WITH-AES-256-GCM-SHA384     TLS-ECDH-ECDSA-WITH-AES-256-CBC-SHA384
 TLS-ECDH-ECDSA-WITH-AES-256-CBC-SHA        TLS-RSA-WITH-AES-256-CCM-8
 TLS-RSA-WITH-CAMELLIA-256-GCM-SHA384       TLS-RSA-WITH-CAMELLIA-256-CBC-SHA256
 TLS-RSA-WITH-CAMELLIA-256-CBC-SHA          TLS-ECDH-RSA-WITH-CAMELLIA-256-GCM-SHA384
 TLS-ECDH-RSA-WITH-CAMELLIA-256-CBC-SHA384  TLS-ECDH-ECDSA-WITH-CAMELLIA-256-GCM-SHA384
 TLS-ECDH-ECDSA-WITH-CAMELLIA-256-CBC-SHA384 TLS-ECDH-ECDSA-WITH-ARIA-256-GCM-SHA384
 TLS-ECDH-RSA-WITH-ARIA-256-GCM-SHA384      TLS-RSA-WITH-ARIA-256-GCM-SHA384
 TLS-ECDH-ECDSA-WITH-ARIA-256-CBC-SHA384    TLS-ECDH-RSA-WITH-ARIA-256-CBC-SHA384
 TLS-RSA-WITH-ARIA-256-CBC-SHA384           TLS-RSA-WITH-AES-128-GCM-SHA256
 TLS-RSA-WITH-AES-128-CCM                   TLS-RSA-WITH-AES-128-CBC-SHA256
 TLS-RSA-WITH-AES-128-CBC-SHA               TLS-ECDH-RSA-WITH-AES-128-GCM-SHA256
 TLS-ECDH-RSA-WITH-AES-128-CBC-SHA256       TLS-ECDH-RSA-WITH-AES-128-CBC-SHA
 TLS-ECDH-ECDSA-WITH-AES-128-GCM-SHA256     TLS-ECDH-ECDSA-WITH-AES-128-CBC-SHA256
 TLS-ECDH-ECDSA-WITH-AES-128-CBC-SHA        TLS-RSA-WITH-AES-128-CCM-8
 TLS-RSA-WITH-CAMELLIA-128-GCM-SHA256       TLS-RSA-WITH-CAMELLIA-128-CBC-SHA256
 TLS-RSA-WITH-CAMELLIA-128-CBC-SHA          TLS-ECDH-RSA-WITH-CAMELLIA-128-GCM-SHA256
 TLS-ECDH-RSA-WITH-CAMELLIA-128-CBC-SHA256  TLS-ECDH-ECDSA-WITH-CAMELLIA-128-GCM-SHA256
 TLS-ECDH-ECDSA-WITH-CAMELLIA-128-CBC-SHA256 TLS-ECDH-ECDSA-WITH-ARIA-128-GCM-SHA256
 TLS-ECDH-RSA-WITH-ARIA-128-GCM-SHA256      TLS-RSA-WITH-ARIA-128-GCM-SHA256
 TLS-ECDH-ECDSA-WITH-ARIA-128-CBC-SHA256    TLS-ECDH-RSA-WITH-ARIA-128-CBC-SHA256
 TLS-RSA-WITH-ARIA-128-CBC-SHA256           TLS-RSA-PSK-WITH-CHACHA20-POLY1305-SHA256
 TLS-RSA-PSK-WITH-AES-256-GCM-SHA384        TLS-RSA-PSK-WITH-AES-256-CBC-SHA384
 TLS-RSA-PSK-WITH-AES-256-CBC-SHA           TLS-RSA-PSK-WITH-CAMELLIA-256-GCM-SHA384
 TLS-RSA-PSK-WITH-CAMELLIA-256-CBC-SHA384   TLS-RSA-PSK-WITH-ARIA-256-GCM-SHA384
 TLS-RSA-PSK-WITH-ARIA-256-CBC-SHA384       TLS-RSA-PSK-WITH-AES-128-GCM-SHA256
 TLS-RSA-PSK-WITH-AES-128-CBC-SHA256        TLS-RSA-PSK-WITH-AES-128-CBC-SHA
 TLS-RSA-PSK-WITH-CAMELLIA-128-GCM-SHA256   TLS-RSA-PSK-WITH-CAMELLIA-128-CBC-SHA256
 TLS-RSA-PSK-WITH-ARIA-128-GCM-SHA256       TLS-RSA-PSK-WITH-ARIA-128-CBC-SHA256
 TLS-PSK-WITH-CHACHA20-POLY1305-SHA256      TLS-PSK-WITH-AES-256-GCM-SHA384
 TLS-PSK-WITH-AES-256-CCM                   TLS-PSK-WITH-AES-256-CBC-SHA384
 TLS-PSK-WITH-AES-256-CBC-SHA               TLS-PSK-WITH-CAMELLIA-256-GCM-SHA384
 TLS-PSK-WITH-CAMELLIA-256-CBC-SHA384       TLS-PSK-WITH-AES-256-CCM-8
 TLS-PSK-WITH-ARIA-256-GCM-SHA384           TLS-PSK-WITH-ARIA-256-CBC-SHA384
 TLS-PSK-WITH-AES-128-GCM-SHA256            TLS-PSK-WITH-AES-128-CCM
 TLS-PSK-WITH-AES-128-CBC-SHA256            TLS-PSK-WITH-AES-128-CBC-SHA
 TLS-PSK-WITH-CAMELLIA-128-GCM-SHA256       TLS-PSK-WITH-CAMELLIA-128-CBC-SHA256
 TLS-PSK-WITH-AES-128-CCM-8                 TLS-PSK-WITH-ARIA-128-GCM-SHA256
 TLS-PSK-WITH-ARIA-128-CBC-SHA256
Last error was: -0xFFFFFFFF - ERROR - Generic error
```

So it seems valuable to constrain this list as well. But, again, our cipher list does need updates.

**But**, why is this failing for you, then? That's one thing that I haven't figured out yet. We build with mbedTLS and it's working fine. I'm assuming that Debian is ripping out bad ciphers themselves, instead of using everything that comes out of the box for mbedTLS. And that there's just nothing in the intersection of the list of ciphers that we request, the list of ciphers that Debian's mbedTLS build supports, and GitHub.com's supported cipher list.

I'll open a PR that updates our cipher list to the intermediate set of defaults, and we can see if that helps.

**Comment by ethomson** [comment]
If you're willing to try out a PR build, I've updated here: https://github.com/libgit2/libgit2/pull/6960

**Comment by ethomson** [comment]
Should be fixed in v1.9, and we'll make a stronger update in v2.0.
